File: src/test/java/it/unibo/crabinv/Model/level/LevelTest.java
package it.unibo.crabinv.Model.level;

import it.unibo.crabinv.Model.entities.enemies.wave.Wave;
import it.unibo.crabinv.Model.entities.enemies.wave.WaveProvider;
import it.unibo.crabinv.Model.entities.enemies.wave.WaveSequence;
import it.unibo.crabinv.Model.levels.Level;
import it.unibo.crabinv.Model.levels.LevelImpl;
import org.junit.jupiter.api.Test;

import java.util.ArrayDeque;
import java.util.List;
import java.util.Queue;

import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.Mockito.mock;

class LevelTest {

    /**
     * WaveProvider deterministico per testare l’avanzamento delle waves.
     */
    private static final class QueueWaveProvider implements WaveProvider {
        private final Queue<Wave> waves;

        private QueueWaveProvider(final List<Wave> waves) {
            this.waves = new ArrayDeque<>(waves);
        }

        @Override
        public boolean hasMoreWaves() {
            return !waves.isEmpty();
        }

        @Override
        public Wave getNextWave() {
            return waves.remove();
        }
    }

    @Test
    void constructorShouldStartFromFirstWave() {
        final Wave w1 = mock(Wave.class);
        final Wave w2 = mock(Wave.class);
        final WaveProvider provider = new QueueWaveProvider(List.of(w1, w2));

        final Level level = new LevelImpl(provider);

        assertSame(w1, level.getCurrentWave(), "Il costruttore deve posizionare il livello sulla prima wave");
        assertFalse(level.isLevelFinished(), "Un livello con wave corrente non deve essere finito");
    }

    @Test
    void advanceWaveShouldMoveToNextWaveAndThenFinish() {
        final Wave w1 = mock(Wave.class);
        final Wave w2 = mock(Wave.class);
        final WaveProvider provider = new QueueWaveProvider(List.of(w1, w2));

        final Level level = new LevelImpl(provider);
        assertSame(w1, level.getCurrentWave());

        level.advanceWave();
        assertSame(w2, level.getCurrentWave(), "advanceWave() deve passare alla wave successiva");
        assertFalse(level.isLevelFinished());

        level.advanceWave();
        assertNull(level.getCurrentWave(), "Finite le wave, la currentWave deve diventare null");
        assertTrue(level.isLevelFinished(), "Con currentWave == null il livello deve risultare finito");

        level.advanceWave();
        assertNull(level.getCurrentWave(), "Dopo la fine del livello, advanceWave() deve mantenere lo stato finito");
        assertTrue(level.isLevelFinished());
    }

    @Test
    void levelWithNoWavesShouldBeImmediatelyFinished() {
        final WaveProvider provider = new QueueWaveProvider(List.of());

        final Level level = new LevelImpl(provider);

        assertNull(level.getCurrentWave(), "Se il provider non ha wave, currentWave deve essere null");
        assertTrue(level.isLevelFinished(), "Se non ci sono wave, il livello deve risultare finito");
    }

    @Test
    void shouldReturnWavesInOrderAndThenReportNoMoreWaves() {
        final Wave w1 = mock(Wave.class);
        final Wave w2 = mock(Wave.class);

        final WaveProvider provider = new WaveSequence(List.of(w1, w2));

        assertTrue(provider.hasMoreWaves());
        assertSame(w1, provider.getNextWave());

        assertTrue(provider.hasMoreWaves());
        assertSame(w2, provider.getNextWave());

        assertFalse(provider.hasMoreWaves());
    }
}

File: src/test/java/it/unibo/crabinv/model/level/LevelTest.kt
package it.unibo.crabinv.model.level

import it.unibo.crabinv.model.entities.enemies.wave.Wave
import it.unibo.crabinv.model.entities.enemies.wave.WaveProvider
import it.unibo.crabinv.model.entities.enemies.wave.WaveSequence
import it.unibo.crabinv.model.levels.Level
import it.unibo.crabinv.model.levels.LevelImpl
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test
import org.mockito.Mockito
import java.util.ArrayDeque
import java.util.Queue

internal class LevelTest {
    @Test
    fun constructorShouldStartFromFirstWave() {
        val w1 = Mockito.mock(Wave::class.java)
        val w2 = Mockito.mock(Wave::class.java)
        val provider: WaveProvider = QueueWaveProvider(listOf<Wave?>(w1, w2) as MutableList<Wave?>)

        val level: Level = LevelImpl(provider)

        Assertions.assertSame(w1, level.currentWave, "The constructor should position itself in the first wave")
        Assertions.assertFalse(level.isLevelFinished, "A level with a wave ongoing shouldn't be finished")
    }

    @Test
    fun advanceWaveShouldMoveToNextWaveAndThenFinish() {
        val w1 = Mockito.mock(Wave::class.java)
        val w2 = Mockito.mock(Wave::class.java)
        val provider: WaveProvider = QueueWaveProvider(listOf<Wave?>(w1, w2) as MutableList<Wave?>)

        val level: Level = LevelImpl(provider)
        Assertions.assertSame(w1, level.currentWave)

        level.advanceWave()
        Assertions.assertSame(w2, level.currentWave, "advanceWave() should go to the next wave")
        Assertions.assertFalse(level.isLevelFinished)

        level.advanceWave()
        Assertions.assertNull(
            level.currentWave,
            "After all waves are finished, the current wave should become null",
        )
        Assertions.assertTrue(level.isLevelFinished, "If currentWave == null the level should be finished")

        level.advanceWave()
        Assertions.assertNull(
            level.currentWave,
            "After level is finished, advanceWave() should maintain a finite status",
        )
        Assertions.assertTrue(level.isLevelFinished)
    }

    @Test
    fun levelWithNoWavesShouldBeImmediatelyFinished() {
        val provider: WaveProvider = QueueWaveProvider(mutableListOf())

        val level: Level = LevelImpl(provider)

        Assertions.assertNull(level.currentWave, "If the provider hasn't got any waves, it should be null")
        Assertions.assertTrue(level.isLevelFinished, "If there aren't any waves, the level should be finished")
    }

    @Test
    fun shouldReturnWavesInOrderAndThenReportNoMoreWaves() {
        val w1 = Mockito.mock(Wave::class.java)
        val w2 = Mockito.mock(Wave::class.java)

        val provider: WaveProvider = WaveSequence(listOf<Wave?>(w1, w2))

        Assertions.assertTrue(provider.hasMoreWaves())
        Assertions.assertSame(w1, provider.getNextWave())

        Assertions.assertTrue(provider.hasMoreWaves())
        Assertions.assertSame(w2, provider.getNextWave())

        Assertions.assertFalse(provider.hasMoreWaves())
    }

    /**
     * WaveProvider used to test the advancement of waves.
     */
    private class QueueWaveProvider(
        waves: MutableList<Wave?>,
    ) : WaveProvider {
        private val waves: Queue<Wave?> = ArrayDeque<Wave?>(waves)

        override fun hasMoreWaves(): Boolean = !waves.isEmpty()

        override fun getAllWaves(): MutableList<Wave?> = mutableListOf()

        override fun getNextWave(): Wave? = waves.remove()
    }
}
